File: platform/lang-impl/src/com/intellij/formatting/commandLine/FormatterStarter.java
/*
 * Copyright 2000-2016 JetBrains s.r.o.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.intellij.formatting.commandLine;

import com.intellij.openapi.application.ApplicationManager;
import com.intellij.openapi.application.ApplicationStarterEx;
import com.intellij.openapi.application.ex.ApplicationEx;
import com.intellij.openapi.application.ex.ApplicationInfoEx;
import com.intellij.openapi.application.impl.ApplicationInfoImpl;
import com.intellij.openapi.diagnostic.Logger;
import com.intellij.openapi.options.SchemeImportException;
import com.intellij.openapi.vfs.VfsUtil;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.codeStyle.CodeStyleSettings;
import com.intellij.psi.impl.source.codeStyle.CodeStyleSettingsLoader;
import org.jetbrains.annotations.NotNull;

import java.io.File;
import java.io.IOException;
import java.io.PrintWriter;

/**
 * A launcher class for command-line formatter.
 */
public class FormatterStarter extends ApplicationStarterEx {

  public static final String FORMAT_COMMAND_NAME = "format";
  private static final Logger LOG = Logger.getInstance("#" + FormatterStarter.class.getName());

  @Override
  public boolean isHeadless() {
    return true;
  }

  @Override
  public String getCommandName() {
    return FORMAT_COMMAND_NAME;
  }

  @Override
  public void premain(String[] args) {
  }

  @Override
  public void main(String[] args) {
    @SuppressWarnings("UseOfSystemOutOrSystemErr")
    MessageOutput messageOutput = new MessageOutput(
      new PrintWriter(System.out),
      new PrintWriter(System.err));
    messageOutput.info(getAppInfo() + " Formatter\n");
    FileSetFormatter fileSetFormatter = new FileSetFormatter(messageOutput);
    logArgs(args);
    if (args.length < 2) {
      showUsageInfo(messageOutput);
    }
    for (int i = 1; i < args.length; i ++) {
      if (args[i].startsWith("-")) {
        if (checkOption(args[i], "-h", "-help")) {
          showUsageInfo(messageOutput);
        }
        if (checkOption(args[i], "-s", "-settings")) {
          //noinspection AssignmentToForLoopParameter
          i ++;
          if (i >= args.length) {
            fatalError(messageOutput, "Missing settings file path.");
          }
          try {
            CodeStyleSettings settings = readSettings(args[i]);
            fileSetFormatter.setCodeStyleSettings(settings);
          }
          catch (SchemeImportException e) {
            fatalError(messageOutput, e.getLocalizedMessage() + "\n");
          }
        }
        else if (checkOption(args[i], "-r", "-R")) {
          fileSetFormatter.setRecursive();
        }
        else if (checkOption(args[i], "-m", "-mask")) {
          //noinspection AssignmentToForLoopParameter
          i ++;
          if (i >= args.length) {
            fatalError(messageOutput, "Missing file mask(s).");
          }
          for (String mask : args[i].split(",")) {
            if (!mask.isEmpty()) {
              fileSetFormatter.addFileMask(mask);
            }
          }
        }
        else {
          fatalError(messageOutput, "Unknown option " + args[i]);
        }
      }
      else {
        try {
          fileSetFormatter.addEntry(args[i]);
        }
        catch (IOException e) {
          fatalError(messageOutput, e.getLocalizedMessage());
        }
      }
    }
    try {
      fileSetFormatter.processFiles();
      messageOutput.info("\n" + fileSetFormatter.getProcessedFiles() + " file(s) formatted.\n");
    }
    catch (IOException e) {
      fatalError(messageOutput, e.getLocalizedMessage());
    }
    ((ApplicationEx)ApplicationManager.getApplication()).exit(true, true);
  }

  private static void fatalError(@NotNull MessageOutput messageOutput, @NotNull String message) {
    messageOutput.error("ERROR: " + message + "\n");
    System.exit(1);
  }

  private static CodeStyleSettings readSettings(@NotNull String settingsPath) throws SchemeImportException {
    VirtualFile vFile = VfsUtil.findFileByIoFile(new File(settingsPath), true);
    CodeStyleSettingsLoader loader = new CodeStyleSettingsLoader();
    if (vFile == null) {
      throw new SchemeImportException("Cannot find file " + settingsPath);
    }
    return loader.loadSettings(vFile);
  }

  private static boolean checkOption(@NotNull String arg, String... variants) {
    for (String variant: variants) {
      if (variant.equals(arg)) return true;
    }
    return false;
  }

  private static String getAppInfo() {
    ApplicationInfoImpl appInfo = (ApplicationInfoImpl)ApplicationInfoEx.getInstanceEx();
    return String.format("%s, build %s", appInfo.getFullApplicationName(), appInfo.getBuild().asString());
  }


  private static void showUsageInfo(@NotNull MessageOutput messageOutput) {
    messageOutput.info("Usage: format [-h] [-r|-R] [-s|-settings settingsPath] path1 path2...\n");
    messageOutput.info("  -h|-help       Show a help message and exit.\n");
    messageOutput.info("  -s|-settings   A path to Intellij IDEA code style settings .xml file.\n");
    messageOutput.info("  -r|-R          Scan directories recursively.\n");
    messageOutput.info("  -m|-mask       A comma-separated list of file masks.\n");
    messageOutput.info("  path<n>        A path to a file or a directory.\n");
    System.exit(0);
  }

  private static void logArgs(@NotNull String[] args) {
    StringBuilder sb = new StringBuilder();
    for (String arg : args) {
      if (sb.length() > 0) sb.append(",");
      sb.append(arg);
    }
    LOG.info("Arguments: " + sb);
  }
}

File: platform/lang-impl/src/com/intellij/formatting/commandLine/FormatterStarter.kt
// Copyright 2000-2022 JetBrains s.r.o. and contributors. Use of this source code is governed by the Apache 2.0 license.
package com.intellij.formatting.commandLine

import com.intellij.application.options.CodeStyle
import com.intellij.formatting.commandLine.CodeStyleProcessorBuildException.ArgumentsException
import com.intellij.formatting.commandLine.CodeStyleProcessorBuildException.ShowUsageException
import com.intellij.ide.impl.OpenProjectTask
import com.intellij.openapi.application.ApplicationInfo
import com.intellij.openapi.application.ApplicationStarter
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.project.ProjectManager
import com.intellij.openapi.project.ex.ProjectManagerEx
import com.intellij.openapi.util.io.FileUtil
import com.intellij.openapi.vfs.VfsUtil
import com.intellij.psi.codeStyle.CodeStyleSettings
import com.intellij.psi.codeStyle.CodeStyleSettingsManager
import com.intellij.psi.impl.source.codeStyle.CodeStyleSettingsLoader
import com.intellij.util.PlatformUtils
import com.intellij.util.application
import org.jetbrains.jps.model.serialization.PathMacroUtil
import java.io.File
import java.io.IOException
import java.nio.charset.Charset
import java.nio.file.Files
import java.util.*

private val LOG = Logger.getInstance(FormatterStarter::class.java)

/**
 * A launcher class for command-line formatter.
 */
internal class FormatterStarter : ApplicationStarter {
  private val messageOutput = StdIoMessageOutput

  override fun main(args: List<String>) {
    messageOutput.info("$appInfo Formatter\n")
    LOG.info(args.joinToString(",", prefix = "Attributes: "))

    val builder = try {
      createBuilder(args, messageOutput)
    }
    catch (e: ShowUsageException) {
      messageOutput.info(usageInfo)
      exit(0)
      return
    }
    catch (e: ArgumentsException) {
      messageOutput.error("ERROR: ${e.message}\n")
      exit(1)
      return
    }

    val projectUID = UUID.randomUUID().toString()
    val project = createProject(projectUID)
    val processor = builder.build(project)

    try {
      processor.let {
        it.processFiles()
        it.printReport()
        if (!it.isResultSuccessful()) {
          exit(1)
          return
        }
      }
    }
    catch (e: IOException) {
      messageOutput.error("ERROR: ${e.localizedMessage}\n")
      exit(1)
      //return
    }
    finally {
      ProjectManager.getInstance().closeAndDispose(project)
    }

    exit(0)
  }
}



fun createBuilder(args: List<String>, messageOutput: MessageOutput = StdIoMessageOutput): CodeStyleProcessorBuilder =
  CodeStyleProcessorBuilder(messageOutput)
    .apply {
      if (args.size < 2) throw ShowUsageException()
      val skipFlag = Skipper()

      args
        .asSequence().drop(1)  // Skip first argument "format" -- routing from the top level

        // try to treat every two arguments as an argument and its param.
        // If param is not needed we'll take it as an arg on the next iteration,
        // otherwise we'll skip using a boolean flag `skipNext`
        .windowed(size = 2, step = 1, partialWindows = true) { Pair(it[0], it.getOrNull(1)) }
        .forEach { (arg, param) ->
          skipFlag.check { return@forEach }

          when (arg) {
            "-h", "-help" -> throw ShowUsageException()
            "-r", "-R" -> recursive()
            "-d", "-dry" -> dryRun()
            "-s", "-settings" -> {
              param ?: throw ArgumentsException("Missing settings file path.")
              withCodeStyleSettings(readSettings(param) ?: throw ArgumentsException("Cannot find file $param"))
              skipFlag.skip()
            }
            "-m", "-mask" -> {
              withFileMasks(param ?: throw ArgumentsException("Missing file mask(s)."))
              skipFlag.skip()
            }
            "-charset" -> {
              param ?: throw ArgumentsException("Missing file mask(s).")
              runCatching { Charset.forName(param) }
                .onSuccess { withCharset(it) }
                .onFailure { messageOutput.error("Ignoring charset setting: ${it.message}") }
              skipFlag.skip()
            }
            "-allowDefaults" -> allowFactoryDefaults()
            else -> {
              if (arg.startsWith("-")) throw ArgumentsException("Unknown option $arg")
              withEntry(arg)
            }
          }
        }
    }


private const val usageInfo = """
Usage: format [-h] [-r|-R] [-d|-dry] [-s|-settings settingsPath] [-charset charsetName] [-allowDefaults] path1 path2...
  -h|-help         Show a help message and exit.
  -s|-settings     A path to Intellij IDEA code style settings .xml file. This setting will be
                   be used as a primary one regardless to the surrounding project settings
  -r|-R            Scan directories recursively.
  -d|-dry          Perform a dry run: no file modifications, only exit status.
  -m|-mask         A comma-separated list of file masks.
  -charset         Force charset to use when reading and writing files.
  -allowDefaults   Use factory defaults when style is not defined for a given file. I.e. when -s
                   is not not set and file doesn't belong to any IDEA project. Otherwise file will
                   be ignored.
  path<n>        A path to a file or a directory.  
"""

fun readSettings(settingsFile: File): CodeStyleSettings? {
  return VfsUtil.findFileByIoFile(settingsFile, true)
    ?.let {
      it.refresh(false, false)
      CodeStyleSettingsLoader().loadSettings(it)
    }
}

private fun readSettings(settingsPath: String): CodeStyleSettings? = readSettings(File(settingsPath))

private val appInfo: String
  get() = ApplicationInfo.getInstance()
    .let { "${it.fullApplicationName}, build ${it.build.asString()}" }

sealed class CodeStyleProcessorBuildException : RuntimeException {
  constructor() : super()
  constructor(message: String) : super(message)

  class ShowUsageException : CodeStyleProcessorBuildException()
  class ArgumentsException(message: String) : CodeStyleProcessorBuildException(message)
}

private class Skipper(private var skip: Boolean = false) {
  fun skip() {
    skip = true
  }

  inline fun check(action: () -> Unit) {
    if (skip) {
      skip = false
      action()
    }
  }
}

private fun exit(code: Int) {
  application.exit(true, true, false, code)
}

private val PROJECT_DIR_PREFIX = PlatformUtils.getPlatformPrefix() + ".format."
private const val PROJECT_DIR_SUFFIX = ".tmp"

private fun createProjectDir(projectUID: String) = FileUtil
  .createTempDirectory(PROJECT_DIR_PREFIX, projectUID + PROJECT_DIR_SUFFIX)
  .toPath()
  .resolve(PathMacroUtil.DIRECTORY_STORE_NAME)
  .also { Files.createDirectories(it) }

private fun createProject(projectUID: String) =
  ProjectManagerEx.getInstanceEx()
    .openProject(createProjectDir(projectUID), OpenProjectTask(isNewProject = true))
    ?.also {
      CodeStyle.setMainProjectSettings(it, CodeStyleSettingsManager.getInstance().createSettings())
    }
  ?: throw RuntimeException("Failed to create temporary project $projectUID")